File: Loenn/entities/energy_booster.lua
local utils = require "utils"
local drawableSprite = require "structs.drawable_sprite"

local energyBooster = {}

energyBooster.name = "vitellary/energybooster"
energyBooster.placements = {
    {
        name = "normal",
        data = {
            behaveLikeDash = false,
            redirectSpeed = false,
            oneUse = false,
        },
    },
    {
        name = "redirect",
        data = {
            behaveLikeDash = false,
            redirectSpeed = true,
            oneUse = false,
        },
    },
}

function energyBooster.sprite(room, entity)
    return drawableSprite.fromTexture(entity.redirectSpeed and "objects/energyBoosterRedirect/booster00" or "objects/energyBooster/booster00", entity)
end

function energyBooster.selection(room, entity)
    return utils.rectangle(entity.x - 10, entity.y - 10, 20, 20)
end

return energyBooster

File: Ahorn/entities/energyBooster.jl
module FlushelineEnergyBooster

using ..Ahorn, Maple

@mapdef Entity "vitellary/energybooster" EnergyBooster(x::Integer, y::Integer, behaveLikeDash::Bool=false, redirectSpeed::Bool=false, oneUse::Bool=false)

const placements = Ahorn.PlacementDict(
    "Energy Booster (Crystalline)" => Ahorn.EntityPlacement(
        EnergyBooster
    ),
    "Energy Booster (Redirect) (Crystalline)" => Ahorn.EntityPlacement(
        EnergyBooster,
        "point",
        Dict{String, Any}(
            "redirectSpeed" => true
        )
    ),
)

sprite = "objects/energyBooster/booster00"
altSprite = "objects/energyBoosterRedirect/booster00"

function Ahorn.selection(entity::EnergyBooster)
    x, y = Ahorn.position(entity)
    redirect = get(entity.data, "redirectSpeed", false)

    return Ahorn.getSpriteRectangle(redirect ? altSprite : sprite, x, y)
end

function Ahorn.render(ctx::Ahorn.Cairo.CairoContext, entity::EnergyBooster, room::Maple.Room)
    redirect = get(entity.data, "redirectSpeed", false)

    Ahorn.drawSprite(ctx, redirect ? altSprite : sprite, 0, 0)
end

end